src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -0,0 +1,152 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Moq;
+using NBitcoin;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Configuration.Logging;
+using Stratis.Bitcoin.Configuration.Settings;
+using Stratis.Bitcoin.Consensus;
+using Stratis.Bitcoin.Utilities;
+using Xunit;
+
+namespace Stratis.Bitcoin.Tests.Consensus
+{
+    public class ChainedHeaderTreeTest
+    {
+        public class TestContext
+        {
+            public Network Network = Network.RegTest;
+            public Mock<IChainedHeaderValidator> ChainedHeaderValidatorMock = new Mock<IChainedHeaderValidator>();
+            public Mock<ICheckpoints> CheckpointsMock = new Mock<ICheckpoints>();
+            public Mock<IChainState> ChainStateMock = new Mock<IChainState>();
+            public ConsensusSettings ConsensusSettings = new ConsensusSettings();
+
+            public ChainedHeaderTree ChainedHeaderTree;
+
+            public ChainedHeaderTree CreateChainedHeaderTree()
+            {
+                this.ChainedHeaderTree = new ChainedHeaderTree(this.Network, new ExtendedLoggerFactory(), this.ChainedHeaderValidatorMock.Object, this.CheckpointsMock.Object, this.ChainStateMock.Object, this.ConsensusSettings);
+                return this.ChainedHeaderTree;
+            }
+
+            public ChainedHeader ExtendAChain(int count, ChainedHeader chainedHeader = null)
+            {
+                ChainedHeader previousHeader = chainedHeader ?? new ChainedHeader(this.Network.GetGenesis().Header, this.Network.GenesisHash, 0);
+
+                foreach (int index in Enumerable.Range(1, count))
+                {
+                    BlockHeader header  = this.Network.Consensus.ConsensusFactory.CreateBlockHeader();
+                    header.HashPrevBlock = previousHeader.HashBlock;
+                    header.Bits = previousHeader.Header.Bits - 1000; // just increase difficulty.
```

Comment:
Could you please expand the comment? I cannot see the logic behind. I think it should be implemented differently, but if you want to go for this shortcut, please at least explain how it works.

Reply:
ok can you suggest how it should be implemented?

Reply:
So 1st what is the problem that this implementation right? If I recall correctly, Genesis difficulty should probably be 1, I'm not sure about that on Stratis network though. But on bitcoin I think it was 1. Difficulty one is coded as `shift+0x00FFFF` and the right part of the difficulty can go down only to `0x008000`and all values below that are invalid. Therefore your for loop here only works if can't is lower than 33. If it is more you will produce invalid values.

So you should actually use some helper methods that will help you do what you want to do instead of directly touching bits which is encoding. Directly touching the base is very hacky way. So for example you could do is work with the difficulty number and increase that and then convert it to the bits.

Another way to do that is to create headers with the correct timestamps and then simply let the difficulty be calculated from that is normal.